File: src/test/java/eu/execom/fabut/AbstractFabutRepositoryAssertTest.java
package eu.execom.fabut;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;

import fabut.model.Contract;
import fabut.model.Employee;
import fabut.model.EmployeeDTO;
import fabut.model.WithListProperty;
import fabut.model.WithMapProperty;

/**
 * TODO add comments
 * 
 * @author Dusko Vesin
 * @author Nikola Olah
 * @author Bojan Babic
 * @author Nikola Trkulja
 */
public abstract class AbstractFabutRepositoryAssertTest extends Assert
		implements IFabutRepositoryTest {
	// mock lists
	private List<Object> entityTierOneTypes = new ArrayList<Object>();
	private List<Object> entityTierTwoTypes = new ArrayList<Object>();
	private List<Object> noDefaultConstructorEntities = new ArrayList<Object>();
	private List<Object> employees = new ArrayList<Object>();
	private FabutRepositoryAssert fabutRepositoryAssert;

	public AbstractFabutRepositoryAssertTest() {

	}

	@Override
	public List<Class<?>> getEntityTypes() {
		final List<Class<?>> entityTypes = new LinkedList<Class<?>>();
		entityTypes.add(Employee.class);
		return entityTypes;
	}

	@Override
	public List<Object> findAll(final Class<?> entityClass) {
		if (entityClass == Employee.class) {
			return employees;
		}
		return null;
	}

	@Override
	public Object findById(final Class<?> entityClass, final Object id) {
		if (entityClass == Employee.class) {
			for (final Object entity : employees) {
				if (id.equals(((Employee) entity).id())) {
					return entity;
				}
			}
		}
		return null;
	}

	public List<Object> getEntityTierOneTypes() {
		return entityTierOneTypes;
	}

	public void setEntityTierOneTypes(final List<Object> list1) {
		entityTierOneTypes = list1;
	}

	public List<Object> getEntityTierTwoTypes() {
		return entityTierTwoTypes;
	}

	public void setEntityTierTwoTypes(final List<Object> list2) {
		entityTierTwoTypes = list2;
	}

	public List<Object> getNoDefaultConstructorEntities() {
		return noDefaultConstructorEntities;
	}

	public void setNoDefaultConstructorEntities(
			final List<Object> noDefaultConstructorEntities) {
		this.noDefaultConstructorEntities = noDefaultConstructorEntities;
	}

	@Override
	@Before
	public void fabutBeforeTest() {
		fabutRepositoryAssert = new FabutRepositoryAssert(this);
	}

	@Override
	@After
	public void fabutAfterTest() {
	}

	@Override
	public List<Class<?>> getComplexTypes() {
		final List<Class<?>> complexTypes = new LinkedList<Class<?>>();
		complexTypes.add(Contract.class);
		complexTypes.add(Technology.class);
		complexTypes.add(EmployeeDTO.class);
		complexTypes.add(WithListProperty.class);
		complexTypes.add(WithMapProperty.class);
		return complexTypes;
	}

	@Override
	public List<Class<?>> getIgnoredTypes() {
		final List<Class<?>> ignoredTypes = new LinkedList<Class<?>>();
		return ignoredTypes;
	}

	@Override
	public void customAssertEquals(final Object expected, final Object actual) {
		Assert.assertEquals(expected, actual);

	}

	public FabutRepositoryAssert getFabutRepositoryAssert() {
		return fabutRepositoryAssert;
	}

	public List<Object> getEmployees() {
		return employees;
	}

	public void setEmployees(List<Object> employees) {
		this.employees = employees;
	}

}

File: src/test/java/eu/execom/fabut/AbstractFabutRepositoryAssertTest.scala
package eu.execom.fabut

import org.junit.Assert
import scala.reflect.runtime.universe.{ Type, typeOf }
import eu.execom.fabut.model.EntityTierOneType
import eu.execom.fabut.model.TierTwoType
import eu.execom.fabut.model.ObjectWithComplexProperty
import eu.execom.fabut.model.TierSixType
import eu.execom.fabut.model.TierOneType
import eu.execom.fabut.model.CopyCaseClass
import eu.execom.fabut.model.ObjectInsideSimpleProperty
import eu.execom.fabut.model.IgnoredType
import eu.execom.fabut.model.ObjectWithSimpleProperties
import eu.execom.fabut.model.TierTwoTypeWithIgnoredType
import eu.execom.fabut.model.ObjectWithMap
import eu.execom.fabut.model.EntityTierThreeType
import eu.execom.fabut.model.Person
import eu.execom.fabut.model.TierThreeType
import eu.execom.fabut.model.ObjectWithSimpleList
import eu.execom.fabut.model.EntityTierTwoType
import eu.execom.fabut.model.ObjectWithSimpleMap
import eu.execom.fabut.model.ListType
import eu.execom.fabut.model.TierFiveType
import eu.execom.fabut.model.EmptyClass
import eu.execom.fabut.model.TierFourType
import eu.execom.fabut.model.TrivialClasses._
import eu.execom.fabut.model.EntityTierOneType
import eu.execom.fabut.model.EntityTierTwoType
import org.junit.Before
import org.junit.After
import eu.execom.fabut.enums.AssertType
import eu.execom.fabut.model.NoDefaultConstructorEntity
import scala.collection.mutable.ListBuffer
import eu.execom.fabut.model.TierOneTypeDuplicate
import eu.execom.fabut.model.test.Student
import eu.execom.fabut.model.test.Faculty
import eu.execom.fabut.model.test.Address
import eu.execom.fabut.model.test.Teacher

class AbstractFabutRepositoryAssertTest extends Assert with IFabutRepositoryTest {

  //mock lists
  var entityTierOneTypes: ListBuffer[EntityTierOneType] = ListBuffer()
  var entityTierTwoTypes: ListBuffer[EntityTierTwoType] = ListBuffer()
  var noDefaultConstructorEntities: ListBuffer[NoDefaultConstructorEntity] = ListBuffer()

  var _fabutRepositoryAssert: FabutRepositoryAssert = null

  @Before
  override def fabutBeforeTest = {
    _fabutRepositoryAssert = new FabutRepositoryAssert(this, AssertType.REPOSITORY_ASSERT)
  }

  @After
  override def fabutAfterTest = {
  }

  def fabutRepositoryAssert = _fabutRepositoryAssert

  override def entityTypes: List[Type] = {

    val entityTypes: ListBuffer[Type] = ListBuffer()
    entityTypes += typeOf[EntityTierOneType]
    entityTypes += typeOf[EntityTierTwoType]
    entityTypes += typeOf[EntityTierThreeType]
    entityTypes += typeOf[NoDefaultConstructorEntity]
    return entityTypes.toList
  }

  override def findAll(clazz: Type) = {

    if (typeOf[EntityTierOneType] == clazz) {
      entityTierOneTypes.toList
    } else if (typeOf[EntityTierTwoType] == clazz) {
      entityTierTwoTypes.toList
    } else {
      noDefaultConstructorEntities.toList
    }

  }

  override def findById(entityClass: Type, id: Any) = {
    if (typeOf[EntityTierOneType] == entityClass) {
      entityTierOneTypes.find(entity => entity.id == id)
    } else if (typeOf[EntityTierTwoType] == entityClass) {
      entityTierTwoTypes.find(entity => entity.id == id)
    } else {
      noDefaultConstructorEntities.find(entity => entity.id == id)
    }
  }

  override def complexTypes: List[Type] = {

    val complexTypes: ListBuffer[Type] = ListBuffer()
    complexTypes += typeOf[ObjectWithSimpleProperties]
    complexTypes += typeOf[EntityTierThreeType]
    complexTypes += typeOf[ObjectWithComplexProperty]
    complexTypes += typeOf[ObjectInsideSimpleProperty]
    complexTypes += typeOf[ObjectWithMap]
    complexTypes += typeOf[ObjectWithSimpleMap]
    complexTypes += typeOf[ObjectWithSimpleList]
    complexTypes += typeOf[CopyCaseClass]
    complexTypes += typeOf[Person]
    complexTypes += typeOf[TierOneType]
    complexTypes += typeOf[TierTwoType]
    complexTypes += typeOf[TierThreeType]
    complexTypes += typeOf[TierFourType]
    complexTypes += typeOf[TierFiveType]
    complexTypes += typeOf[TierSixType]
    complexTypes += typeOf[TierTwoTypeWithIgnoredType]
    complexTypes += typeOf[EmptyClass]
    complexTypes += typeOf[ListType]
    complexTypes += typeOf[TierOneTypeDuplicate]
    complexTypes += typeOf[A]
    complexTypes += typeOf[B]
    complexTypes += typeOf[C]
    complexTypes += typeOf[D]
    complexTypes += typeOf[E]
    complexTypes += typeOf[Student]
    complexTypes += typeOf[Faculty]
    complexTypes += typeOf[Address]
    complexTypes += typeOf[Teacher]

    return complexTypes.toList
  }

  override def ignoredTypes: List[Type] = {
    val ignoredTypes: ListBuffer[Type] = ListBuffer()
    ignoredTypes += typeOf[IgnoredType]
    return ignoredTypes.toList
  }

  override def customAssertEquals(expectedObject: Any, actualObject: Any) {
    Assert.assertEquals(expectedObject, actualObject)
  }

  def setEntityTierOneTypes(entityTierOneTypes1: ListBuffer[EntityTierOneType]) {
    entityTierOneTypes = entityTierOneTypes1
  }

  def setEntityTierTwoTypes(entityTierTwoTypes1: ListBuffer[EntityTierTwoType]) {
    entityTierTwoTypes = entityTierTwoTypes1
  }

}